encodings/sequence/src/array.rs

```diff
@@ -0,0 +1,292 @@
+use num_traits::cast::FromPrimitive;
+use vortex_array::arrays::PrimitiveArray;
+use vortex_array::stats::{ArrayStats, StatsSetRef};
+use vortex_array::vtable::{
+    ArrayVTable, CanonicalVTable, NotSupported, OperationsVTable, VTable, ValidityVTable,
+    VisitorVTable,
+};
+use vortex_array::{
+    ArrayBufferVisitor, ArrayChildVisitor, ArrayRef, Canonical, EncodingId, EncodingRef, vtable,
+};
+use vortex_dtype::Nullability::NonNullable;
+use vortex_dtype::{DType, NativePType, PType, match_each_integer_ptype, match_each_native_ptype};
+use vortex_error::{VortexExpect, VortexResult, vortex_bail, vortex_err};
+use vortex_mask::Mask;
+use vortex_scalar::{PValue, Scalar, ScalarValue};
+
+vtable!(Sequence);
+
+#[derive(Clone, Debug)]
+/// An array representing the equation `A[i] = base + i * multiplier`.
+pub struct SequenceArray {
+    base: PValue,
+    multiplier: PValue,
+    dtype: DType,
+    length: usize,
+    stats_set: ArrayStats,
+}
+
+impl SequenceArray {
+    pub fn typed_new<T: NativePType + Into<PValue>>(
+        base: T,
+        multiplier: T,
+        length: usize,
+    ) -> VortexResult<Self> {
+        Self::new(base.into(), multiplier.into(), T::PTYPE, length)
+    }
+
+    // Constructs a sequence array using two integer values (with the same ptype).
+    pub fn new(
+        base: PValue,
+        multiplier: PValue,
+        ptype: PType,
+        length: usize,
+    ) -> VortexResult<Self> {
+        if !ptype.is_int() {
+            vortex_bail!("only integer ptype are supported in SequenceArray currently")
+        }
+
+        Self::try_last(base, multiplier, ptype, length).map_err(|e| {
+            e.with_context(format!(
+                "final value not expressible, base = {:?}, multiplier = {:?}, len = {} ",
+                base, multiplier, length
+            ))
+        })?;
+
+        Ok(Self::unchecked_new(base, multiplier, ptype, length))
+    }
+
+    pub(crate) fn unchecked_new(
+        base: PValue,
+        multiplier: PValue,
+        ptype: PType,
+        length: usize,
+    ) -> Self {
+        let dtype = DType::Primitive(ptype, NonNullable);
+        Self {
+            base,
+            multiplier,
+            dtype,
+            length,
+            // TODO(joe): add stats, on construct or on use?
+            stats_set: Default::default(),
+        }
+    }
+
+    pub fn ptype(&self) -> PType {
+        self.dtype.as_ptype()
+    }
+
+    pub fn base(&self) -> PValue {
+        self.base
+    }
+
+    pub fn multiplier(&self) -> PValue {
+        self.multiplier
+    }
+
+    fn try_last(
+        base: PValue,
+        multiplier: PValue,
+        ptype: PType,
+        length: usize,
+    ) -> VortexResult<PValue> {
+        match_each_integer_ptype!(ptype, |$P| {
+            let len_t = <$P>::from_usize(length-1)
+                .ok_or_else(|| vortex_err!("cannot convert length {} into {}", length, ptype))?;
+
+            let base = <$P>::try_from(base)?;
```

Comment:
there's a slightly nicer syntax you can use now with `let base = base.as_primitive::<$P>()?;`, same for multiplier. Here and on line 114-115 and 168-169